### DISCUSSION TITLE
Reproduce detectCommunity

### DISCUSSION THREAD

**Discussion by ychen983384** [networkit/networkit#1227] [Q&A]
Runs on the same graph generates different results. Is there a way to make this reproducible on the same graph?
# first run
Communities = nk.community.detectCommunities(GData, algo=nk.community.PLM(G=GData,refine=True, gamma=0.5))
Communities detected in 1.21094 [s]
solution properties:
-------------------  -------------
communities            10
min community size    12562
max community size    41579
avg. community size   24544.8
imbalance                 1.69399
edge cut             404216
edge cut (portion)        0.054895
modularity                0.801823

# second run
Communities = nk.community.detectCommunities(GData, algo=nk.community.PLM(G=GData,refine=True, gamma=0.5))
Communities detected in 1.18332 [s]
solution properties:
-------------------  --------------
communities            11
min community size    13536
max community size    43623
avg. community size   22313.5
imbalance                 1.95496
edge cut             426229
edge cut (portion)        0.0578845
modularity                0.808974

**Comment by clstaudt** [comment]
@ychen983384  Have you tried turning off parallelism?

```python
PLM(G, par="none")
```

**Comment by ychen983384** [comment]
> @ychen983384 Have you tried turning off parallelism?
> 
> ```python
> PLM(G, par="none")
> ```
@clstaudt 
I have not tried that. The reason I tried networkit's PLM is because I have a large graph to work with (~15 million vertices) and hope the parallelism could shorten the runnig time for detect community. I would test turning off parallelism. Do you suggest that the results would not be reproducible with parallelism on?

**Comment by fabratu** [comment]
Yes, with parallelism the order of execution (in terms of nodes moving from one cluster to the other) is not deterministic. Without parallelism you get the same result for consequetive runs:

```
import networkit as nk

G = nk.generators.BarabasiAlbertGenerator(5,100).generate()

for par_strat in ["none", "balanced"]:
    PLM = nk.community.PLM(G,par=par_strat)
    PLM.run()
    par1 = PLM.getPartition().getVector()

    PLM = nk.community.PLM(G,par=par_strat)
    PLM.run()
    par2 = PLM.getPartition().getVector()

    print(par1 == par2)
```

The output is `True, False`.

**Comment by clstaudt** [comment]
@ychen983384  As @fabratu has explained, the parallelism makes it nondeterministic. Turning off parallelism for your graph kind of defeats the purpose of using this particular algorithm. We never considered determinism to be an important property when building it.

**Comment by ychen983384** [comment]
> @ychen983384 As @fabratu has explained, the parallelism makes it nondeterministic. Turning off parallelism for your graph kind of defeats the purpose of using this particular algorithm. We never considered determinism to be an important property when building it.

@clstaudt @fabratu  Thank you both for the clear explanation and the great tools you developed. Scientific community and journals have been emphasizing more and more on reproducibility of both computational and experimental research work which might make determinism more important. With determinism, the parallel tools you have developed such as PLM and parLeiden would be more attractive to the scientific community since huge data set come from single cell multi-omics technology. Is it feasible to implement these community detection algorithms (such as louvain and leiden) with both parallelism and determinism or it is theoretically impossible?

**Reply by clstaudt** [reply]
> Is it feasible to implement these community detection algorithms (such as louvain and leiden) with both parallelism and determinism or it is theoretically impossible?

Probably not theoretically impossible. But what would be the motivation to invest into developing such a version? ->

> Scientific community and journals have been emphasizing more and more on reproducibility of both computational and experimental research work which might make determinism more important. With determinism, the parallel tools you have developed such as PLM and parLeiden would be more attractive to the scientific community since huge data set come from single cell multi-omics technology.

So the difference between detecting 10 or 11 clusters makes a major difference for the research results? And if we could apply some trick (similar to fixing a random seed) to get determinism, so that we fix the result to 11 clusters, that would make the research reproducible, while otherwise it isn't? I have trouble following the argument.

**Reply by ychen983384** [reply]
In my view, reproducibility means the same input+computational environment+running script -> the exact same output (total cluster number and cluster assignment of each node/vertice). Many tools provide an option for a fixed random seed to achieve that as you mentioned. I noted networkit also has a option to set the seed (networkit.engineering.setSeed(seed, useThreadId)) but there is a separate random number generator per thread. Not sure whether that is related with the non-determinism.

**Reply by hmeyerhenke** [reply]
While I acknowledge the need, a solution is not so easy. The non-determinism results from the parallelism itself, since it can lead to different orders in which certain computations are made. This is different from randomization. To change that, one would probably have to restrict the parallelism to certain execution orders. Not impossible, but probably not so easy. Since our resources are quite limited these days, I don't see this coming any time soon (out of our group) -- unless the NetworKit community jumps in and provides such an implementation.

**Reply by clstaudt** [reply]
@ychen983384  Okay, I understand your definition of reproducibility, I'd call it determinism. I'd like to understand why that is a priority or even requirement for the scientific community you mentioned. 

"For the findings of a study to be reproducible means that results obtained by an experiment or an observational study or in a statistical analysis of a data set should be achieved again with a high degree of reliability when the study is replicated." - Wikipedia definition of reproducibility in science

If you ensure determinism of the experiment code, does that imply that your research is reproducible (in the Wikipedia sense)? If there are some statistical variations in the output between experiment runs, does it mean your research is not reproducible?

**Reply by ychen983384** [reply]
@clstaudt That's my understanding for reproducibility or in your term of determinism that other people can replicate the analysis to get the exact same results with the same dataset and script. Currently, scientific journals usually ask authors to share the raw data and the analysis scripts so that people can replicate the analysis done in the study.
I am from single cell omics field and we usually perform cell clustering  based on their gene expression  and annotate them as different cell types. If each time the run gives different membership assignment for each cell, their cell type annotation would be different. I would not say that some statistical variations of replicates is not reproducible as long as the scientific conclusion does not change. Some times we also test runs with different random seeds to ensure our conclusion is robust.
@hmeyerhenke Thanks for your explanation and totally understand that.

**Comment by ilan-gold** [comment]
> This is different from randomization. To change that, one would probably have to restrict the parallelism to certain execution orders. Not impossible, but probably not so easy. Since our resources are quite limited these days, I don't see this coming any time soon (out of our group) -- unless the NetworKit community jumps in and provides such an implementation.

I'm in the same boat from the same field and have the same question.  I am wondering if this is even technically possible. It sounds like it is possible, though, to fix the execution order given an RNG-like input.  Do I have that right? I'm only interested in `leiden` at the moment
